staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go

```diff
@@ -0,0 +1,264 @@
+	restclient "k8s.io/client-go/rest"
+	cache "k8s.io/client-go/tools/cache"
+	workqueue "k8s.io/client-go/util/workqueue"
+
+	rmtypesv1a1 "k8s.io/api/flowcontrol/v1alpha1"
+
+	"k8s.io/klog"
+)
+
+// This request filter implements https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+
+// FairQueuingFactory knows how to make FairQueueingSystem objects.
+// This filter makes a FairQueuingSystem for each priority level.
+type FairQueuingFactory interface {
+	NewFairQueuingSystem(concurrencyLimit, numQueues, queueLengthLimit int, requestWaitLimit time.Duration, clk clock.Clock) FairQueuingSystem
+}
+
+// FairQueuingSystem is the abstraction for the queuing and
+// dispatching functionality of one non-exempt priority level.  It
+// covers the functionality described in the "Assignment to a Queue",
+// "Queuing", and "Dispatching" sections of
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+// .  Some day we may have connections between priority levels, but
+// today is not that day.
+type FairQueuingSystem interface {
+
+	// The following methods modify or query the configuration
+
+	SetConcurrencyLimit(int)
+	SetDesiredNumQueues(int)
+	GetDesiredNumQueues() int
+	GetActualNumQueues() int
+	SetQueueLengthLimit(int)
+	SetRequestWaitLimit(time.Duration)
+
+	// DoOnEmpty records a function to be called when all the queues
+	// have no requests waiting nor executing.  The filter uses this
+	// for a priority level that has become undesired, setting a
+	// callback that will actually remove the priority level by
+	// establishing a new RMState whose priorityLevelStates map lacks
+	// that entry.  If the priority level becomes desired again before
+	// this is called, the filter cancels by calling `DoOnEmpty(nil)`.
+	// The callback is invoked with no mutexes locked.
+	DoOnEmpty(func())
+
+	// Wait, in the happy case, shuffle shards the given request into
+	// a queue and eventually dispatches the request from that queue.
+	// Dispatching means to return with `execute==true`.  In the
+	// unhappy cases the request is eventually rejected, which means
+	// to return with `execute=false`.  In the happy case the caller
+	// is required to invoke the returned `afterExecution` after the
+	// request is done executing.  The hash value and hand size are
+	// used to do the shuffle sharding.
+	Wait(hashValue uint64, handSize int32) (execute bool, afterExecution func())
+}
+
+// RMState is the variable state that this filter is working with at a
+// given point in time.
+type RMState struct {
+	// flowSchemas holds the flow schema objects, sorted by increasing
+	// numerical (decreasing logical) matching precedence
+	flowSchemas FlowSchemaSeq
+
+	// priorityLevelStates maps the PriorityLevelConfiguration object
+	// name to the state for that level
+	priorityLevelStates map[string]*PriorityLevelState
```

Comment:
it's better if we use uid as key here? avoiding PL w/ the same name got removed and created instantly?

Reply:
Yes, that would be the implementation if we agree on that behavior.  We have not agreed on what the desired behavior is.  https://github.com/kubernetes/enhancements/pull/1098/files#r300193486